src/Features/Core/Portable/Workspace/CompileTimeSolutionProvider.cs

```diff
@@ -44,20 +45,18 @@ public Factory()
         private const string RazorSourceGeneratorTypeName = "Microsoft.NET.Sdk.Razor.SourceGenerators.RazorSourceGenerator";
         private static readonly string s_razorSourceGeneratorFileNamePrefix = Path.Combine(RazorSourceGeneratorAssemblyName, RazorSourceGeneratorTypeName);
 
-        private readonly Workspace _workspace;
-
         private readonly object _gate = new();
 
         /// <summary>
-        /// Cached compile time solution corresponding to the <see cref="Workspace.PrimaryBranchId"/>
+        /// Cached compile-time solution corresponding to an existing design-time solution.
         /// </summary>
-        private (int DesignTimeSolutionVersion, BranchId DesignTimeSolutionBranch, Solution CompileTimeSolution)? _primaryBranchCompileTimeCache;
+#if NETCOREAPP
+        private readonly ConditionalWeakTable<Solution, Solution> _designTimeToCompileTimeSolution = new();
+#else
+        private ConditionalWeakTable<Solution, Solution> _designTimeToCompileTimeSolution = new();
+#endif
 
-        /// <summary>
-        /// Cached compile time solution for a forked branch.  This is used primarily by LSP cases where
-        /// we fork the workspace solution and request diagnostics for the forked solution.
-        /// </summary>
-        private (int DesignTimeSolutionVersion, BranchId DesignTimeSolutionBranch, Solution CompileTimeSolution)? _forkedBranchCompileTimeCache;
```

Comment:
@dibarbet i believe this has the semantics from before.  Previously we keyed basd on Version+BranchId.  But these change every time you get a new solution.  So the only time these woudl match would be if you passed in teh exact same (reference-equals) Solution instance from before.  So we can get the same behavior by using a CWT keyed off Solution instances.

Reply:
The idea here being that as soon as no one is using the solution we can get rid of the design time solution.  Makes sense to me.